src/impl_/pymodule.rs

```diff
@@ -74,9 +82,22 @@ impl ModuleDef {
         let module = unsafe {
             Py::<PyModule>::from_owned_ptr_or_err(py, ffi::PyModule_Create(self.ffi_def.get()))?
         };
-        if self.initialized.swap(true, atomic::Ordering::SeqCst) {
+        let current_interpreter =
+            unsafe { ffi::PyInterpreterState_GetID(ffi::PyInterpreterState_Get()) };
+        let initialized_interpreter = py.allow_threads(|| {
+            #[cfg(once_lock)]
+            {
+                *self.interpreter.get_or_init(|| current_interpreter)
+            }
+
+            #[cfg(not(once_lock))]
+            {
+                *self.interpreter.lock().get_or_insert(current_interpreter)
+            }
+        });
+        if current_interpreter != initialized_interpreter {
             return Err(PyImportError::new_err(
-                "PyO3 modules may only be initialized once per interpreter process",
+                "PyO3 modules do not yet support subinterpreters, see https://github.com/PyO3/pyo3/issues/576",
             ));
         }
         (self.initializer.0)(py, module.as_ref(py))?;
```

Comment:
Do we really want to call the initializer multiple times or should the second import become a no-op?

Reply:
I think the spirit of PEP 489 (which we don't support yet, but hopefully are heading towards) is that modules can be torn down and re-imported, so I think it's correct to call the initializer multiple times.

Reply:
> which we don't support yet, but hopefully are heading towards

I think this is the main issue here, isn't it? Does this change mean we opt into allowing tear-down and reinitialization of PyO3-based modules? What invariants would this break? Do we test this somewhere?

(I am not trying to say that it doesn't work. Just that I myself do understand the consequences of this change and little confidence about its effects on existing code.)

Reply:
In https://github.com/pydantic/pydantic/issues/6584#issuecomment-1673260886, you mention opt-in. Might this be advisable even for this limited relaxation? What if the module is torn down but static data still contains pointers into the old module (e.g. capsules) which are now invalid? For example, is the code in https://github.com/PyO3/rust-numpy/blob/c16fbb1e630b0538fdf17cbc53043bf0467459f9/src/npyffi/mod.rs#L30-L32 valid after this change?

Reply:
So I found the Cython implementation here: 

https://github.com/cython/cython/blob/ce1aa59ab6300e18e50c752b5c9e28a04b665060/Cython/Utility/ModuleSetupCode.c#L1648C18-L1648C18

That code:
 - Has a similar check that the interpreter hasn't changed ID
 - Caches a module instance in a C static and returns the single module instance always.

So I guess let's implement this strategy too and defer the question of PEP 489 compatibility to a future PR. My goal for this PR is primarily to fix the use cases in the OP, which I believe the Cython strategy would cover.

Reply:
While I think this strategy is significantly safer, I am still hesitant. While this should resolve keeping static data pointing into the modules valid, what about the other way around, i.e. the module containing data bound to the interpreter? Is it guaranteed that finalizing and re-initializing a single interpreter in the same process will yield a new interpreter ID? (I guess we lack the API to actually test ATM?)

Reply:
That's a great question. I might be able to craft a test using `with_embedded_interpreter`, will try tomorrow.

Reply:
TLDR: finalizing and re-initializing yields the same ID, Cython is similarly bugged here, and [`Py_Finalize`]https://docs.python.org/3/c-api/init.html#c.Py_FinalizeEx specifically calls out that things aren't generally expected to work more than once. (I made `with_embedded_python_interpreter` unsafe for this reason.)

It's clearly awkward to get this right. I think on balance it's desirable if we match Cython so that more of the ecosystem fails in the same way and it seems better that we only ever run the module initializer once for now. We can make progress towards per-module state to eventually remove this worry.

---

The interpreter ID is not incremented. For example, this code doesn't hit our safety check (the interpreter ID is 0 both times):

```rust
use pyo3::{prelude::*, wrap_pymodule};

#[pymodule]
fn my_module(_py: Python, _m: &PyModule) -> PyResult<()> {
    Ok(())
}

#[test]
fn test_reloading_interpreter_doesnt_allow_module() {
    unsafe {
        pyo3::with_embedded_python_interpreter(|py| wrap_pymodule!(my_module)(py));

        pyo3::with_embedded_python_interpreter(|py| wrap_pymodule!(my_module)(py));
    }
}
```

I tried a variant of this with a Cython module, and I get a crash which looks like it's related to extension types being incorrectly reused across interpreters:

```
#[test]
fn test_reloading_interpreter_doesnt_allow_module() {
    unsafe {
        pyo3::with_embedded_python_interpreter(|py| {
            py.import("yaml._yaml").unwrap();
        });

        pyo3::with_embedded_python_interpreter(|py| {
            py.import("yaml._yaml").map_err(|e| e.print(py)).unwrap();
        });
    }
}
```

Reply:
> It's clearly awkward to get this right. I think on balance it's desirable if we match Cython so that more of the ecosystem fails in the same way and it seems better that we only ever run the module initializer once for now. We can make progress towards per-module state to eventually remove this worry.

I agree with that. Do you know the upstream position on this, i.e. would they consider fixing this or does it work as intended?

> (I made with_embedded_python_interpreter unsafe for this reason.)

I am somewhat confused as to what the consequences of this discovery are. Especially with the Cython module immediately showing the failure mode I was concerned about. Can we still relax this? Is the `unsafe` on `with_embedded_python_interpreter` sufficient? (I guess another other way to re-initialize Python in the current process would call unsafe Rust code or code in inherently unsafe languages like C, i.e. is trusted?)

Reply:
I have just posted https://github.com/python/cpython/issues/109785 - let's see what the outcome of that discussion is.

Reply:
No movement yet on the upstream issue. @adamreichold are you happy if I proceed to merge this? As above at least we'll then be consistent with Cython and I think the generally accepted position is that things are expected to be wonky after finalizing and re-initializing the interpreter.

Reply:
Yes, I think the language on `with_embedded_python_interpreter` is strong enough. This is the only API we provide to finalize and then re-initialize the interpreter?

Reply:
Yes (apart from the raw FFI bindings). `auto-initialize` never bothers finalizing the interpreter.